Allow to chain NativePromise::ThenCommand even if it returned void

jyavenard opened the pull request:
#### ab34eaa17f5d07b2b3ce0c8d5e3af57f9c29c8a3
<pre>
Allow to chain NativePromise::ThenCommand even if it returned void
<a href="https://bugs.webkit.org/show_bug.cgi?id=265306">https://bugs.webkit.org/show_bug.cgi?id=265306</a>
<a href="https://rdar.apple.com/118770512">rdar://118770512</a>

Reviewed by Youenn Fablet.

When the then/whenSettled handler runs and it returned void, handle it as if a resolved GenericPromise was returned.

Add API test.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::myMethodReturningThenCommandWithPromise):
(TestWebKitAPI::myMethodReturningThenCommandWithVoid):
(TestWebKitAPI::TEST):
(TestWebKitAPI::myMethodReturningThenCommand): Deleted.

Canonical link: <a href="https://commits.webkit.org/271267@main">https://commits.webkit.org/271267@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f04d1398c7cb96457c288c3f89e1c8e731359940

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27865 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6504 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29125 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30399 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25435 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28359 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8473 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3898 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28131 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5252 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23902 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/4539 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4718 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24925 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31087 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/24156 "Found 1 new JSC stress test failure: wasm.yaml/wasm/gc/js-api.js.wasm-collect-continuously (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25459 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25368 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/27020 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4731 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2885 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6251 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/34496 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6687 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5178 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7455 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5198 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -792,42 +789,40 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
+            {
+                Locker lock { m_lock };
+                completionProducer = std::exchange(m_completionProducer, { });
+            }
```
Suggestion, while you're touching this line: Instead of a default init + assignment, how about an immediately-invoked lambda? Something like `auto cP = [this]() { lock; return exchange(...); }();`

squelart reviewed (commented):
Unofficial review, nice simplifying changes!
One optional suggestion.

jyavenard committed:
Allow to chain NativePromise::ThenCommand even if it returned void
https://bugs.webkit.org/show_bug.cgi?id=265306
rdar://118770512

Reviewed by Youenn Fablet.

When the then/whenSettled handler runs and it returned void, handle it as if a resolved GenericPromise was returned.

Add API test.

* Source/WTF/wtf/NativePromise.h:
* Tools/TestWebKitAPI/Tests/WTF/NativePromise.cpp:
(TestWebKitAPI::myMethodReturningThenCommandWithPromise):
(TestWebKitAPI::myMethodReturningThenCommandWithVoid):
(TestWebKitAPI::TEST):
(TestWebKitAPI::myMethodReturningThenCommand): Deleted.

Canonical link: https://commits.webkit.org/271267@main
